### PR TITLE
Add go implementation

### PR DESCRIPTION
**Pull request by lucas-s-work**
Adds:
 - Go implementation
 - Docker config
 - CI go config

### CODE REVIEW

File: go/views/index.pug [hunk 1]
```diff
@@ -0,0 +1,9 @@
+head
+  title #{.title}
+  script.
+    function redirect(){
+      window.location.href = "/redirect"
+    }
+body
+  p #{.message}
+  button(style="font-size:20px",onclick="redirect()") Start journey!
```
Comment (hatstand) [inline go/views/index.pug:9]: Can't this just be a link rather than using a script?

File: .circleci/config.yml [hunk 2]
```diff
@@ -21,8 +21,50 @@ jobs:
           working_directory: javascript
           command: docker build -t creditkudos-development.com/auth-demo .
 
+  go_build:
```
Comment (hatstand) [inline .circleci/config.yml:24]: Can you rename this to reflect that it's building a docker image not a go binary?

File: go/main.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	clientID     = ""
+	clientSecret = ""
+)
+
+func main() {
+	redirectURL = os.Getenv("REDIRECTURL")
+	clientID = os.Getenv("CLIENTID")
+	clientSecret = os.Getenv("CLIENTSECRET")
+	if redirectURL == "" || clientID == "" || clientSecret == "" {
+		panic("Set env variables before running")
```
Comment (hatstand) [inline go/main.go:31]: Use `log.Fatal()` instead

File: go/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,11 @@
+FROM golang:1.14
+  WORKDIR /home/app/repo
+  COPY . /home/app/repo
+
+  ENV CLIENTID=
```
Comment (hatstand) [inline go/Dockerfile:5]: Underscores in these for consistency, e.g., `CLIENT_ID`.

File: go/main.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+	engine := pug.New("./views", ".pug")
+	app := fiber.New(&fiber.Settings{Views: engine})
+	oauthClient := createOauthClient()
+
+	createRoutes(app, oauthClient)
+
+	if err := app.Listen(8080); err != nil {
+		fmt.Println("App errored: ", err.Error())
```
Comment (hatstand) [inline go/main.go:41]: `log.Fatal`

File: go/main.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+		panic("Set env variables before running")
+	}
+
+	engine := pug.New("./views", ".pug")
+	app := fiber.New(&fiber.Settings{Views: engine})
+	oauthClient := createOauthClient()
+
+	createRoutes(app, oauthClient)
+
+	if err := app.Listen(8080); err != nil {
```
Comment (hatstand) [inline go/main.go:40]: Let's use `3000` for consistency with the other sample and print out where it's listening.

File: go/main.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+	// Setup landing page
+	app.Get("/", func(c *fiber.Ctx) {
+		if err := c.Render(
+			"index",
+			fiber.Map{
+				"title":   "Example Go implementation",
+				"message": "Click below to start a journey",
+			},
+		); err != nil {
+			fmt.Println("Failed to load main page:", err)
```
Comment (hatstand) [inline go/main.go:68]: `c.SendStatus(http.StatusInternalServerError)` too?

File: go/main.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+		tokenString, err := token.SignedString([]byte(clientSecret))
+		if err != nil {
+			fmt.Println("Failed to create jwt token:", err)
+			return
+		}
+
+		// Generate the redirection URI and redirect
+		customerToken := oauth2.SetAuthURLParam("customer_token", tokenString)
+		url := oauthConfig.AuthCodeURL("abc", oauth2.AccessTypeOnline, customerToken)
+		fmt.Println("Redirect URL:", url)
```
Comment (hatstand) [inline go/main.go:97]: `log`

File: go/main.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+		c.Redirect(url)
+	})
+
+	// Setup return callback
+	app.Get("/callback", func(c *fiber.Ctx) {
+		code := c.Query("code")
+		fmt.Println("Response code:", code)
```
Comment (hatstand) [inline go/main.go:104]: `log`

File: go/main.go [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+		ctx := c.Context()
+		token, err := oauthConfig.Exchange(ctx, code)
+		if err != nil {
+			handleError(fmt.Errorf("unable to exchange for token: %s", err), c)
+			return
+		}
+
+		fmt.Println("Access Token:", token.AccessToken)
```
Comment (hatstand) [inline go/main.go:112]: `log`

File: go/main.go [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+	// Setup on complete
+	app.Get("/complete", func(c *fiber.Ctx) {
+		if err := c.Render(
+			"index",
+			fiber.Map{
+				"title":   "Complete!",
+				"message": "Journey complete! Click below to start another",
+			},
+		); err != nil {
+			handleError(fmt.Errorf("unable to render /complete: %s", err), c)
```
Comment (hatstand) [inline go/main.go:129]: `%w`

File: go/main.go [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+		}
+	})
+}
+
+// Handle errors and render error response
+func handleError(e error, c *fiber.Ctx) {
+	fmt.Println(e.Error())
```
Comment (hatstand) [inline go/main.go:136]: `log`

File: go/main.go [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+	if err := c.Render("error", fiber.Map{"message": e.Error()}); err != nil {
+		panic(err)
```
Comment (hatstand) [inline go/main.go:138]: `log.Fatal`

File: go/main.go (hunk 6 above)
Comment (lucass-work) [reply go/main.go:40]: When fiber launches it prints out the port. I was unable to get 3000 working in dev as it was `already in use` however 8080 was working.

Comment (hatstand) [reply go/main.go:40]: `3000` will work fine if you're not running other stuff.

File: go/main.go (hunk 7 above)
Comment (lucass-work) [reply go/main.go:68]: Changed to use `handleError`, will add this to `handleError` on fatal.

File: go/views/index.pug [hunk 14]
```diff
@@ -0,0 +1,5 @@
+head
+  title #{.title}
+body
+  p #{.message}
+  button(style="font-size:20px",onclick="location.href='/redirect'") Start journey!
```
Comment (hatstand) [inline go/views/index.pug:5]: Why `onclick` when you can use `<a href=`?

Comment (lucass-work) [reply go/views/index.pug:5]: I liked the idea of it being a button over just a link

Comment (hatstand) [reply go/views/index.pug:5]: If you wrap the button with an `<a>` then it's both.

Comment (lucass-work) [reply go/views/index.pug:5]: Are they not functionally equivalent?

Comment (hatstand) [reply go/views/index.pug:5]: No, as `onclick` is not equivalent accessibility wise.

**Review by hatstand — APPROVED** [review]

**Pull request merged**